hs-bindgen-runtime/src/HsBindgen/Runtime/ByteArray.hs

```diff
@@ -0,0 +1,90 @@
+-- an FFI implementation based on 'Storable' is never going to be /extremely/
+-- fast, as we are effectively (de)serializing. A few additional @memcpy@
+-- operations are therefore not going to be a huge difference.
+--
+-- We /could/ choose to use pinned bytearrays. This would avoid /some/ copying,
+-- but by no means all: we'd still need one copy (instead of two) in
+-- 'peekByteArray' and 'pokeByteArray', and the calls to 'peek' and 'poke' in
+-- 'peekFromByteArray' and 'pokeToByteArray' will (likely) do copying of their
+-- own as well.
+module HsBindgen.Runtime.ByteArray (
+     -- * Support for defining 'Storable' instances for union types
+     peekByteArray
+   , pokeByteArray
+     -- * Support for defining setters and getters for union types
+   , setUnionPayload
+   , getUnionPayload
+   ) where
+
+import Control.Monad.Primitive (RealWorld)
+import Data.Coerce (Coercible, coerce)
+import Data.Primitive.ByteArray (ByteArray, MutableByteArray, copyByteArray, newPinnedByteArray, freezeByteArray, withMutableByteArrayContents, sizeofByteArray)
+import Foreign (Storable (poke, peek), Ptr, castPtr, copyBytes, sizeOf)
+import System.IO.Unsafe (unsafePerformIO)
+
+{-------------------------------------------------------------------------------
+  Support for defining 'Storable' instances for union types
+-------------------------------------------------------------------------------}
+
+peekByteArray :: Ptr a -> Int -> IO ByteArray
+peekByteArray src n = do
+    pinnedCopy <- newPinnedByteArray n
+    withMutableByteArrayContents pinnedCopy $ \dest ->
+      copyBytes dest (castPtr src) n
+    freezeByteArray pinnedCopy 0 n
+
+pokeByteArray :: Ptr a -> ByteArray -> IO ()
+pokeByteArray dest bytes = do
+    pinnedCopy <- thawPinned bytes
+    withMutableByteArrayContents pinnedCopy $ \src ->
+      copyBytes dest (castPtr src) n
+  where
+    n = sizeofByteArray bytes
+
+{-------------------------------------------------------------------------------
+  Support for defining setters and getters for union types
+-------------------------------------------------------------------------------}
+
+setUnionPayload :: forall payload union.
+     ( Storable payload
+     , Storable union
+     , Coercible union ByteArray
+     )
+  => payload -> union
+setUnionPayload = coerce . pokeToByteArray (sizeOf (undefined :: union))
+
+getUnionPayload :: forall payload union.
+     ( Storable payload
+     , Storable union
+     , Coercible union ByteArray
+     )
+  => union -> payload
+getUnionPayload = peekFromByteArray . coerce
+
+peekFromByteArray :: Storable a => ByteArray -> a
+peekFromByteArray bytes = unsafePerformIO $ do
+    pinnedCopy <- thawPinned bytes
+    withMutableByteArrayContents pinnedCopy $ \ptr ->
+      peek (castPtr ptr)
+
+pokeToByteArray :: Storable a => Int -> a -> ByteArray
+pokeToByteArray n x = unsafePerformIO $ do
+    pinnedCopy <- newPinnedByteArray n
```

Comment:
is `n` different from `sizeOf (undefined :: a)`?

Reply:
I guess it can, but it should be always larger.

Reply:
I don't feel terribly strongly, feel free to change it if you prefer.

Reply:
> The reason I didn't make this

The reason is it's usage in `setUnionPayload`, where `union`  (i.e. a) can be larger than `payload`. I think someone should write a comment, as even the author is confused...

Reply:
Oh, hah, you're right. Yes. Comment would be useful :)